core/src/main/scala/kafka/coordinator/group/DelayedJoin.scala

```diff
@@ -33,11 +33,15 @@ import scala.math.{max, min}
  */
 private[group] class DelayedJoin(coordinator: GroupCoordinator,
                                  group: GroupMetadata,
-                                 rebalanceTimeout: Long) extends DelayedOperation(rebalanceTimeout, Some(group.lock)) {
+                                 rebalanceTimeout: Long) extends DelayedOperation(rebalanceTimeout, None) {
 
-  override def tryComplete(): Boolean = coordinator.tryCompleteJoin(group, forceComplete _)
-  override def onExpiration() = coordinator.onExpireJoin()
-  override def onComplete() = coordinator.onCompleteJoin(group)
+  /**
+   * It controls the lock manually since GroupCoordinator#onCompleteJoin() invoked by onComplete() can't be within a
+   * group lock since GroupCoordinator#onCompleteJoin() tries to complete delayed requests.
+   */
+  override def tryComplete(): Boolean = if (group.inLock(group.hasAllMembersJoined)) forceComplete() else false
+  override def onExpiration(): Unit = coordinator.onExpireJoin()
+  override def onComplete(): Unit = coordinator.onCompleteJoin(group)
```

Comment:
This is another case of deadlock. ```DelayedJoin#tryComplete``` is in a **group lock** and it tries to complete other delayed joins which related to same **__consumer_offsets** partition.